### PR TITLE
GC: Update snapshot cache policy to only support enable/disable, not custom values

### PR DESCRIPTION
**Pull request by markfields**
[AB#1362](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/1362)

We have already discussed that we need to move Snapshot Caching code into the Runtime layer.  Until then, rather than attempting to support custom policies in the driver as to max cache duration, we should just hardcode the max to be 2 days and document this as a requirement for GC to work properly.

### CODE REVIEW

File: packages/runtime/container-runtime/src/test/garbageCollection.spec.ts [hunk 1]
```diff
@@ -86,6 +86,8 @@ describe("Garbage Collection Tests", () => {
     // The default GC data returned by `getGCData` on which GC is run. Update this to update the referenced graph.
     const defaultGCData: IGarbageCollectionData = { gcNodes: {} };
 
+    const defaultSnapshotCacheExpiryMs = 2 * 24 * 60 * 60 * 1000;
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/test/garbageCollection.spec.ts:89]: This number is everywhere, I wonder if it's smart to export it from the garbageCollection class.

**Review by tyler-cai-microsoft — APPROVED** [review]
LGTM, just some minor comments

File: packages/loader/container-loader/src/container.ts [hunk 2]
```diff
@@ -662,20 +662,12 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
             this.connectionStateHandler.containerSaved();
         });
 
-        this._deltaManager = this.createDeltaManager();
```
Comment (tyler-cai-microsoft) [inline packages/loader/container-loader/src/container.ts]: Why is this line deleted?

Comment (markfields) [reply packages/loader/container-loader/src/container.ts]: Uhhhhhhhhhhhhhhh

Comment (markfields) [reply packages/loader/container-loader/src/container.ts]: GitHub.com bug/feature I think.  Try refreshing the page.  I merged main and this file is no longer changed.

File: packages/loader/container-loader/src/container.ts [hunk 3]
```diff
@@ -662,20 +662,12 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
...
-                        return new BlobOnlyStorage(this.loader.services.detachedBlobStorage, this.mc.logger);
-                    }
-                    this.mc.logger.sendTelemetryEvent({
-                        eventName: "NoRealStorageInDetachedContainer",
-                    });
-                    throw new Error("Real storage calls not allowed in Unattached container");
-                }
-                return this.storageService;
-            },
+        this.storageService = new ContainerStorageAdapter(
```
Comment (tyler-cai-microsoft) [inline packages/loader/container-loader/src/container.ts:665]: I don't understand why this change is relevant to the current PR.

File: packages/loader/container-loader/src/container.ts (hunk 2 above)
Comment (tyler-cai-microsoft) [reply packages/loader/container-loader/src/container.ts]: I think I looked at just the changes between your last changes and your most recent changes.

File: packages/test/test-utils/src/testObjectProvider.ts [hunk 4]
```diff
@@ -185,9 +185,15 @@ export class EventAndErrorTrackingLogger extends TelemetryLogger {
...
         return {
             expectedNotFound,
-            unexpectedErrors,
+            unexpectedErrors: this.excludeAllowedErrors(unexpectedErrors),
         };
     }
+
+    private excludeAllowedErrors(errors: ITelemetryBaseEvent[]) {
+        return errors.filter((event) => !(event.eventName in [
+            "NoRealStorageInDetachedContainer", // Removed in current version, but still present in previous versions
```
Comment (markfields) [inline packages/test/test-utils/src/testObjectProvider.ts:194]: This is not ideal.  But there are hundreds of failures hitting this so I don't think it's reasonable to update all tests to expect this.

To clarify my comment in the code here, this error event does not exist in main anymore.  But it does in the old compat versions.  It was removed because it's harmless and unhelpful (the code throws, callers can decide if it's a problem or not).

@anthony-murphy - Any tips or feedback on this?  I do have an alternate approach that avoid the codepath hitting this error altogether, but it's not as correct in terms of encapsulation of layers IMO.

Comment (markfields) [reply packages/test/test-utils/src/testObjectProvider.ts]: (I see that `x in array` doesn't work like I thought. Updated in later iteration)

File: packages/common/driver-definitions/src/storage.ts [hunk 5]
```diff
@@ -101,9 +102,16 @@ export interface IDocumentStorageServicePolicies {
...
+     * @deprecated Use snapshotCacheDisabledForTesting instead - Only enabling/disabling will be exposed
      * This policy tells the runtime that the driver will not use cached snapshots older than this value.
      */
     readonly maximumCacheDurationMs?: number;
+
+    /**
+     * If true, the driver will not use the snapshot cache but will always fetch from the service. Defaults to false.
+     * This configuration is only expected to be used for test scenarios.
+     */
+    readonly snapshotCacheDisabledForTesting?: boolean;
```
Comment (agarwal-navin) [inline packages/common/driver-definitions/src/storage.ts:114]: I don't think we should expose this option here in a public interface. We don't want this to be a policy that hosts can set.
The idea behind adding `maximumCacheDurationMs` as a policy was this:
A host can choose to expire their snapshots very quickly in order to be up-to-date. However, they cannot set this value to be greater than 2 days, only smaller. This gives GC an upper bound on the time it needs to wait for sweeping objects.

A proposal of what we can do - Either have `maximumCacheDurationMs` or remove it. Irrespective of that, the maximum duration of 2 days is fixed and enforeced and that is what is communicated to the runtime and GC. If we want to override this value, it can be done via feature flags and that value is communicated to the runtime by the driver.

Comment (agarwal-navin) [reply packages/common/driver-definitions/src/storage.ts:114]: Note that ODSP does not expose these policies today but other drivers do and they can be set by hosts.

Comment (agarwal-navin) [reply packages/common/driver-definitions/src/storage.ts:114]: @vladsud I would love your thoughts on this since I don't have a lot of experience with driver / host poclicies.

Comment (markfields) [reply packages/common/driver-definitions/src/storage.ts:114]: Biggest question from this thread is what policies are implemented by host v. Driver. This policy type is not passed in as a means to control the driver but is exposed by the driver to enable proper interop.

Comment (markfields) [reply packages/common/driver-definitions/src/storage.ts:114]: > If we want to override this value, it can be done via feature flags and that value is communicated to the runtime by the driver.

This is exactly what I'm doing, except I'm using a boolean feature flag not a number, to cement the fact that this 2-day value really isn't configurable.

Maybe we should look through the code together to see if I'm missing something.

Comment (markfields) [reply packages/common/driver-definitions/src/storage.ts:114]: To expand on my previous comment about driver v. host policies:

> I don't think we should expose this option here in a public interface. We don't want this to be a policy that hosts can set.

This policy is set by the driver, not passed in as a means to control the driver. The driver exposes this info as FYI to other layers to enable proper interop.

> The idea behind adding maximumCacheDurationMs as a policy was this:
> A host can choose to expire their snapshots very quickly in order to be up-to-date. However, they cannot set this value to be greater than 2 days, only smaller. 

I wouldn't say it like that.  Hosts can certainly implement their cache to keep older than 2 days.  But the _driver_ declares it will not use cache entries older than this policy - i.e. 2 days.

The problem was, this value was not plumbed through to the place that implemented it.  See #11168.

The change in this PR is to change the language the driver uses to talk about this.  Instead of declaring how old of snapshots may be used, that info is hardcoded throughout the driver, in accordance with the documented requirement of GC.  However, I do want the driver to own whether snapshot cache is disabled altogether, because we need this capability to test GC without waiting for days.

Alternative is to get rid of this policy, and have the Runtime use the same feature flag name, but I would rather that all live within the driver - especially when thinking of other drivers and how they would be implemented.

Comment (agarwal-navin) [reply packages/common/driver-definitions/src/storage.ts:114]: > This policy is set by the driver, not passed in as a means to control the driver. The driver exposes this info as FYI to other layers to enable proper interop.
> 

This is true only for ODSP today. Other drivers do allow you to provide this value during creation - https://github.com/microsoft/FluidFramework/blob/215770d4613df9c3660eea24bb490cd1ed002aec/packages/drivers/routerlicious-driver/src/documentStorageService.ts#L73.

Comment (agarwal-navin) [reply packages/common/driver-definitions/src/storage.ts:114]: > > If we want to override this value, it can be done via feature flags and that value is communicated to the runtime by the driver.
> 
> This is exactly what I'm doing, except I'm using a boolean feature flag not a number, to cement the fact that this 2-day value really isn't configurable.
> 
> Maybe we should look through the code together to see if I'm missing something.

What I was trying to say is do this via feature flags without exposing is as a policy.

Comment (markfields) [reply packages/common/driver-definitions/src/storage.ts:114]: > This is true only for ODSP today. Other drivers do allow you to provide this value during creation 

Oh... good to know, thanks for pointing it out. I'll update the comment I added to that interface in this PR.

> What I was trying to say is do this via feature flags without exposing is as a policy.

This would mean we'd have the ODSP driver and runtime reading the same feature flag and depending on that fact.  Other drivers would not necessarily respect that same flag.  It's a pretty undesirable and brittle coupling.  Besides that or this policy, I don't know how to plumb the data through.

In light of what you pointed out about how this policy interface may be used, I agree it's a little strange to put on there, but I prefer that approach to trying to keep feature flag names / semantics in touch across all driver implementations + runtime.

Comment (markfields) [reply packages/common/driver-definitions/src/storage.ts:114]: @agarwal-navin are you able to explain again your concern with changing the policy? It doesn't feel disruptive to do so, but it seems like it does to you. Or risky somehow?

Comment (markfields) [reply packages/common/driver-definitions/src/storage.ts:114]: Another idea which started as a joke but now I'm thinking about it - what about changing the type of the existing policy to 0 | 2 days (numerically)? That will document the 2 day assumption/requirement but also allow for test override.

Comment (agarwal-navin) [reply packages/common/driver-definitions/src/storage.ts:114]: My concern is that we should be careful about policies since these are meant to be provided by the hosts to the driver. We should not add test specific policies for example. Since we are updating policies here, we should be confident that we are adding the right one and not make the same mistake we did with the existing policy for cache expiry.

Comment (agarwal-navin) [reply packages/common/driver-definitions/src/storage.ts:114]: > Another idea which started as a joke but now I'm thinking about it - what about changing the type of the existing policy to 0 | 2 days (numerically)? That will document the 2 day assumption/requirement but also allow for test override.

This seems similar to making this a boolean and having the cache expiry value as 2 when the policy is true :-). The current policy also works for testing by setting it to 0 right? I am just trying to understand the value of changing this w.r.t. what currently exists (which I agree is not ideal).

Comment (markfields) [reply packages/common/driver-definitions/src/storage.ts:114]: Value is not having to plumb the number throughout the ODSP driver. Can just read the feature flag both places.

Comment (markfields) [reply packages/common/driver-definitions/src/storage.ts:114]: And less surface area in GC to think about differing values over lifetime of a file. 0 is clearly about testing so more lenient to closing the container of it doesn't match, or just letting it be mismatched

Comment (agarwal-navin) [reply packages/common/driver-definitions/src/storage.ts:114]: That sounds good to me.

Comment (markfields) [reply packages/common/driver-definitions/src/storage.ts:114]: Ok. Let's sleep on it. And @vladsud if you can make your way through this thread and give input I'd like to hear it. 

Downside to this plan is it's a breaking change, but practically it's fine for the next few weeks to stage it.

Comment (markfields) [reply packages/common/driver-definitions/src/storage.ts:114]: For context for anyone following along, the reason the existing approach is challenging is (A) it's hard with current code to plumb the same value to both epochTracker and Runtime - but that is solvable - and (B) GC would require the value to never change. So it needs to be persisted in the file. But there's no good way to do that as-is. So the short-term plan is to lock it at 2 days in the code until we get to the long term plan which is to move the code that retrieves snapshots from the cache into the runtime layer. Then we can persist the policy in the file and reject anything from the cache violating that policy. Need more specific design, and need to get it on the backlog. Not an urgent priority in that it doesn't block roll-out of GC to Bohemia.

Comment (markfields) [reply packages/common/driver-definitions/src/storage.ts:114]: #12183

**Comment by markfields** [comment]
New plan, documented (internally) [here](https://microsoft.sharepoint-df.com/:fl:/s/54e45b67-b090-4b62-ba82-a7be9987bef7/EZahQHhp6SdOo8ry90BuEC0BTKr6W1T80O6vTZ-he0p6_A?e=3hi981&nav=cz0lMkZzaXRlcyUyRjU0ZTQ1YjY3LWIwOTAtNGI2Mi1iYTgyLWE3YmU5OTg3YmVmNyZkPWIhdUVoeU4yRk9jRUNzOFhfeVNLSUxCWlJWaFZXc2p2OUlobENZWkp6SFoyeHFkclhtVUVxMlJiVG51Zm9oT0RQNSZmPTAxVlBNQkw0RVdVRkFIUTJQSkU1SEtIU1hTNjVBRzRFQk4mYz0lMkYmZmx1aWQ9MSZhPUxvb3BBcHAmcD0lNDBmbHVpZHglMkZsb29wLXBhZ2UtY29udGFpbmVyJng9JTdCJTIydyUyMiUzQSUyMlQwUlRVSHh0YVdOeWIzTnZablF1YzJoaGNtVndiMmx1ZEMxa1ppNWpiMjE4WWlGMVJXaDVUakpHVDJORlEzTTRXRjk1VTB0SlRFSmFVbFpvVmxkemFuWTVTV2hzUTFsYVNucElXako0Y1dSeVdHMVZSWEV5VW1KVWJuVm1iMmhQUkZBMWZEQXhWbEJOUWt3MFFrMWFRa2RXTWtnME0wdFdSVmxMUlZVM1ZsRllNamRHTlV3JTNEJTIyJTJDJTIyaSUyMiUzQSUyMjgwMmI0ZDdhLTBmMTktNDIzMy04YmUxLTFiNTcxMDcyODljNCUyMiU3RA%3D%3D).
